### PR TITLE
Additional crypto cleanups, add scrypt impl

### PR DESCRIPTION
**Pull request by jasnell**
More reorganization of crypto impl... start consolidating key parts of the Node.js crypto and web crypto impls starting with HKDF and PBKDF2. Add scrypt impl

### CODE REVIEW

File: src/workerd/api/crypto/scrypt.c++ [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+namespace workerd::api {
+
+kj::Maybe<kj::Array<kj::byte>> scrypt(size_t length,
+                                      uint32_t N,
+                                      uint32_t r,
+                                      uint32_t p,
+                                      uint32_t maxmem,
+                                      kj::ArrayPtr<const kj::byte> pass,
+                                      kj::ArrayPtr<const kj::byte> salt) {
+  auto buf = kj::heapArray<kj::byte>(length);
```
Comment (fhanau) [inline src/workerd/api/crypto/scrypt.c++:18]: Node also checks the size of pass and salt here (probably not needed) and applies some error handling by examining the last error if present after calling `EVP_PBE_scrypt()` (which we should adopt too).

File: src/node/internal/crypto_scrypt.ts [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+  N: 16384,
+  r: 8,
+  p: 1,
+  maxmem: 32 << 20,
+};
+
+function validateParameters(password: ArrayLike, salt: ArrayLike,
+                            keylen: number, options: ScryptOptions) : ValidatedScryptOptions {
+  // TODO(soon): Add support for KeyObject input.
+  password = validateByteSource(password, 'password');
```
Comment (fhanau) [inline src/node/internal/crypto_scrypt.ts:87]: Node uses getArrayBufferOrView for password and salt instead (not sure if it makes a difference).

**Review by fhanau — COMMENTED** [review]
LGTM otherwise, let's just make sure we have some error handling here.

File: src/workerd/api/crypto/scrypt.c++ (hunk 1 above)
Comment (jasnell) [reply src/workerd/api/crypto/scrypt.c++:18]: Length check will happen on the ts side.

Comment (fhanau) [reply src/workerd/api/crypto/scrypt.c++:19]: LGTM, I must have missed the length checks before. Error handling works too.

**Review by fhanau — APPROVED** [review]

**Pull request merged**